controlplane/kubeadm/internal/workload_cluster_coredns.go

```diff
@@ -0,0 +1,312 @@
+	if _, ok := migration.Versions[fmt.Sprintf("%d.%d.%d", from.Major, from.Minor, from.Patch)]; !ok {
+		return fmt.Errorf("fromVersion %q is not a compatible coredns version", from.String())
+	}
+	return nil
+}
+
+// getCoreDNSInfo gets
+// - the current CoreDNS image version installed
+// - the current Corefile Configuration of CoreDNS
+// - the coredns deployment
+func (w *Workload) getCoreDNSInfo(ctx context.Context) (*coreDNSInfo, error) {
+	// Get the coredns configmap and corefile.
+	corednsKey := ctrlclient.ObjectKey{Name: coreDNSKey, Namespace: metav1.NamespaceSystem}
+	cm := &corev1.ConfigMap{}
+	if err := w.Client.Get(ctx, corednsKey, cm); err != nil {
+		return nil, errors.Wrapf(err, "error getting %s config map from target cluster", corednsKey.String())
+	}
+	corefile, ok := cm.Data[corefileKey]
+	if !ok {
+		return nil, errors.New("unable to find the CoreDNS Corefile data")
+	}
+
+	// Get the current CoreDNS deployment.
+	deployment := &appsv1.Deployment{}
+	if err := w.Client.Get(ctx, corednsKey, deployment); err != nil {
+		return nil, errors.Wrapf(err, "unable to get %s deployment from target cluster", corednsKey.String())
+	}
+
+	// Parse deployment image.
+	addonImage := deployment.Spec.Template.Spec.Containers[0].Image
+	ref, err := reference.ParseNormalizedNamed(addonImage)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to parse %s deployment image", corednsKey.String())
+	}
+	tagged, hasTag := ref.(reference.Tagged)
+	if !hasTag {
+		return nil, errors.Errorf("%s deployment does not have a valid image tag", corednsKey.String())
+	}
+	parsedTag, err := parseImageTag(tagged.Tag())
+	if err != nil {
+		return nil, err
+	}
+
+	return &coreDNSInfo{
+		Corefile:       corefile,
+		Deployment:     deployment,
+		CurrentVersion: parsedTag,
+	}, nil
+}
+
+// UpdateCoreDNSDeployment will patch the deployment image to the
+// imageRepo:imageTag in the KCP dns. It will also ensure the volume of the
+// deployment uses the Corefile key of the coredns configmap.
+func (w *Workload) updateCoreDNSDeployment(ctx context.Context, info *coreDNSInfo, dns *kubeadmv1.DNS) error {
+	containers := info.Deployment.Spec.Template.Spec.Containers
+	if len(containers) == 0 {
+		return errors.New("failed to update coredns deployment: deployment spec has no containers")
+	}
+
+	// Parse deployment image.
+	image := info.Deployment.Spec.Template.Spec.Containers[0].Image
+	imageRef, err := reference.ParseNormalizedNamed(image)
+	if err != nil {
+		return errors.Wrapf(err, "unable to parse %q deployment image", image)
+	}
+
+	// Handle imageRepository.
+	imageRepository := fmt.Sprintf("%s/%s", reference.Domain(imageRef), reference.Path(imageRef))
+	if dns.ImageRepository != "" {
+		imageRepository = dns.ImageRepository
+	}
+
+	// Handle imageTag.
+	imageRefTag, ok := imageRef.(reference.Tagged)
+	if !ok {
+		return errors.Errorf("failed to update coredns deployment: does not have a valid image tag: %q", image)
+	}
+	imageTag := imageRefTag.Tag()
```

Comment:
I'm wondering if this would read a little less redundant if `coreDNSInfo` was updated to return the output of `imageRefTag.Tag()` for `CurrentVersion`, that way it wouldn't need to be re-parsed here.

It would require that it's current use be preceeded by a call to `parseImageTag`, but we are already calling that on the version to compare against in `updateDNSCorefile` and `validateCoreDNSImageTag` is parsing to semver before doing anything with the value anyway.

Reply:
Refactored this, ptal